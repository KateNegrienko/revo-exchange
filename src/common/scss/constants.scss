$success: #a8e36c;
$error: #ff7a7a;

$black: #1a1c1f;
$grey: #8b959e;
$white: #fff;
$neutral: #ddd;
$ligth-neutral: #f4f4f4;
$primary: #0075eb;
$primary-hover: lighten($primary, 2%);
$primary-pressed: darken($primary, 2%);
$primary-disabled: $grey;

$elevation-1: 0 4px 8px 0 rgba(16, 23, 40, 0.04);

$border-radius: 5px;
$border-radius-m: 10px;
$border-radius-l: 30px;

$space-s: 5px;
$space-m: 10px;
$space-l: 20px;
$space-xl: 40px;
$space-xxl: 80px;

$breakpoints-screen: (
  "mobile": 440px,
  "tablet": 768px,
  "desktop": 1120px,
  "desktop-large": 1440px,
);

$font-button: (
  "desktop": 20px,
  "tablet": 19px,
  "mobile": 18px,
);
