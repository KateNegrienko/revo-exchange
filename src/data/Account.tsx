export interface Account {
    id: string;
    currency: string;
    value: number;
}