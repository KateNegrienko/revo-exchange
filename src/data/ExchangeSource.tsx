export interface IExchangeSource {
  price: number;
  sourceRate: number;
  destinationRate: number;
}
