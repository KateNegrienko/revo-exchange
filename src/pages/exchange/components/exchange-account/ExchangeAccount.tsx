import { FC, useCallback, useState } from "react";
import { Input, Icon, Select } from "semantic-ui-react";

import Card from "../../../../common/card/Card";
import {
  ExchangeAccountType,
  IExchangeAccountProps,
} from "./ExchangeAccount.interface";
import theme from "./ExchangeAccount.module.scss";

const ExchangeAccount: FC<IExchangeAccountProps> = ({
  account,
  rates,
  price,
  type,
  onChangePrice,
  onChangeAccount,
}) => {
  const [inputError, setInputError] = useState("");
  const handleChangeValidation = useCallback(
    (value: number) => {
      if (value > account.value) {
        setInputError(
          "you cannot change more money than there is in your account"
        );
        return false;
      }
      setInputError("");
      return true;
    },
    [account.value]
  );

  const handleChange = useCallback(
    (e) => {
      e.persist();
      const float = parseFloat(e.target.value);
      handleChangeValidation(e.target.value);
      onChangePrice(type, float.toFixed(2));
    },
    [onChangePrice, handleChangeValidation, type]
  );
  const handleChangeAccount = useCallback(
    (e, data) => {
      onChangeAccount(type, data.value);
    },
    [onChangeAccount, type]
  );
  return (
    <Card className={theme.root}>
      <div className={theme.error}>{inputError}</div>
      <div className={theme.row}>
        <Select
          name={account.id}
          className={theme.select}
          size="small"
          onChange={handleChangeAccount}
          value={account.id}
          options={rates}
        />

        <Input
          iconPosition="left"
          type="number"
          size="small"
          className={theme.input}
        >
          <Icon name={type === ExchangeAccountType.SOURCE ? "minus" : "plus"} />
          <input
            type="number"
            value={price}
            step=".01"
            pattern="^\d*(\.\d{0,2})?$"
            max={account.value}
            onChange={handleChange}
          ></input>
        </Input>
      </div>

      <div className={theme.hint}> Balance: {account.value}</div>
    </Card>
  );
};

export default ExchangeAccount;
