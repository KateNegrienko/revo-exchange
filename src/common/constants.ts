import { IBankAccount } from "../data/Account";

export enum CURRENCIES {
  USD = "USD",
  EUR = "EUR",
  GBP = "GBP",
}

export const DEFAULT_PRICES = {
  destinationPrice: "",
  sourcePrice: "",
};

export const INITIAL_ACCOUNTS: IBankAccount[] = [
  {
    id: CURRENCIES.EUR,
    value: 50.3,
    symbol: "€",
  },
  {
    id: CURRENCIES.GBP,
    value: 100,
    symbol: "£",
  },
  {
    id: CURRENCIES.USD,
    value: 50,
    symbol: "$",
  },
];
