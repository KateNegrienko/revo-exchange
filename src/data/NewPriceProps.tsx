import { ExchangeAccountType } from "../pages/exchange/components/exchange-account/ExchangeAccount.interface";
import { IBankAccount } from "./Account";
import { IRate } from "./Rate";

export interface INewPriceProps {
  rates: IRate[];
  type: ExchangeAccountType;
  price: number;
  sourceAccount: IBankAccount;
  destinationAccount: IBankAccount;
}
