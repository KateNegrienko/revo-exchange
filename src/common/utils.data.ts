import { IRate } from "../data/Rate";
import { CURRENCIES, INITIAL_ACCOUNTS } from "./constants";

export const TEST_RATES: IRate[] = [
  {
    key: CURRENCIES.USD,
    value: CURRENCIES.USD,
    text: CURRENCIES.USD,
    price: 1,
  },
  {
    key: CURRENCIES.EUR,
    value: CURRENCIES.EUR,
    text: CURRENCIES.EUR,
    price: 2,
  },
  {
    key: CURRENCIES.GBP,
    value: CURRENCIES.GBP,
    text: CURRENCIES.GBP,
    price: 3,
  },
];

export const utilsPriceTestData = [
  {
    sourceData: {
      price: 10,
      sourceRate: 2,
      destinationRate: 3,
    },
    expectedDataForSource: {
      destinationPrice: "6.67",
      sourcePrice: "10",
    },
    expectedDataForDestination: {
      destinationPrice: "10",
      sourcePrice: "15",
    },
  },
  {
    sourceData: {
      price: 4,
      sourceRate: 1,
      destinationRate: 3,
    },
    expectedDataForSource: {
      destinationPrice: "1.34",
      sourcePrice: "4",
    },
    expectedDataForDestination: {
      destinationPrice: "4",
      sourcePrice: "12",
    },
  },
];

export const utilsSetPriceTestData = [
  {
    newPriceProps: {
      rates: TEST_RATES,
      price: 10,
      sourceAccount: INITIAL_ACCOUNTS[0],
      destinationAccount: INITIAL_ACCOUNTS[1],
    },
    expectedDataForSource: {
      destinationPrice: "6.67",
      sourcePrice: "10",
    },
    expectedDataForDestination: {
      destinationPrice: "10",
      sourcePrice: "15",
    },
  },
  {
    newPriceProps: {
      rates: TEST_RATES,
      price: 4,
      sourceAccount: INITIAL_ACCOUNTS[2],
      destinationAccount: INITIAL_ACCOUNTS[0],
    },
    expectedDataForSource: {
      destinationPrice: "2",
      sourcePrice: "4",
    },
    expectedDataForDestination: {
      destinationPrice: "4",
      sourcePrice: "8",
    },
  },
];
