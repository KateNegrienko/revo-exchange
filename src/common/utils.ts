import { IBankAccount } from "../data/Account";
import { IExchangeSource } from "../data/ExchangeSource";
import { INewPriceProps } from "../data/NewPriceProps";
import { IRate } from "../data/Rate";
import { ExchangeAccountType } from "../pages/exchange/components/exchange-account/ExchangeAccount.interface";
import { DataState } from "../reducers/account/account.types";
import { DEFAULT_PRICES } from "./constants";

export const round = (value: number) => {
  return Math.floor(value * 100) / 100;
};

export const validation = (
  type: ExchangeAccountType,
  balance: number,
  money: number
): string => {
  return money > balance && type === ExchangeAccountType.SOURCE
    ? "You cannot change more money than there is in your account"
    : "";
};

export const exchangeMapping = (
  item: IBankAccount,
  state: DataState
): IBankAccount => {
  switch (item.id) {
    case state.sourceAccount.id:
      return {
        ...item,
        value: round(item.value - Number(state.sourcePrice)),
      };

    case state.destinationAccount.id:
      return {
        ...item,
        value: round(item.value + Number(state.destinationPrice)),
      };

    default:
      return item;
  }
};

export const findRate = (rates: IRate[], rateId: string) => {
  return rates.find(({ key }: IRate) => key === rateId)?.price;
};

export const getPriceWithSourceType = ({
  price,
  sourceRate,
  destinationRate,
}: IExchangeSource) => {
  const destination = round((price / sourceRate) * destinationRate);
  return {
    destinationPrice: destination.toString(),
    sourcePrice: price.toString(),
  };
};

export const getPriceWithDestinationType = ({
  price,
  sourceRate,
  destinationRate,
}: IExchangeSource) => {
  const source = round((price / destinationRate) * sourceRate);

  return {
    destinationPrice: price.toString(),
    sourcePrice: source.toString(),
  };
};

export const calculatePrice = ({
  rates,
  focusInput,
  price,
  sourceAccount,
  destinationAccount,
}: INewPriceProps) => {
  const sourceRate = findRate(rates, sourceAccount.id);
  const destinationRate = findRate(rates, destinationAccount.id);

  if (sourceRate && destinationRate && price > 0) {
    switch (focusInput) {
      case ExchangeAccountType.SOURCE:
        return getPriceWithSourceType({ price, sourceRate, destinationRate });

      case ExchangeAccountType.DESTINATION:
        return getPriceWithDestinationType({
          price,
          sourceRate,
          destinationRate,
        });
    }
  }
  return DEFAULT_PRICES;
};
