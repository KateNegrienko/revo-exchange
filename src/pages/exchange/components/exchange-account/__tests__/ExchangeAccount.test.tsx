import React from "react";
import { render, shallow, ShallowWrapper } from "enzyme";
import ExchangeAccount from "../ExchangeAccount";
import { INITIAL_ACCOUNTS } from "../../../../../common/constants";
import { TEST_RATES } from "../../../../../common/utils.data";
import { ExchangeAccountType } from "../ExchangeAccount.interface";

describe("[Exchange account component]", () => {
  let wrapper: ShallowWrapper;
  const price = "50";
  const onChangePrice = jest.fn();
  const onChangeAccount = jest.fn();

  beforeEach(() => {
    wrapper = shallow(
      <ExchangeAccount
        account={INITIAL_ACCOUNTS[0]}
        rates={TEST_RATES}
        price={price}
        type={ExchangeAccountType.SOURCE}
        onChangePrice={onChangePrice}
        onChangeAccount={onChangeAccount}
      />
    );
  });

  it("[Exchange account snapshot] Should render properly with only required props ", () => {
    const onChangePrice = (e: any) => {};
    const onChangeAccount = () => {};
    expect(
      render(
        <ExchangeAccount
          account={INITIAL_ACCOUNTS[0]}
          rates={TEST_RATES}
          price=""
          type={ExchangeAccountType.SOURCE}
          onChangePrice={onChangePrice}
          onChangeAccount={onChangeAccount}
        />
      )
    ).toMatchSnapshot();
  });

  it("[input] Should render input with correct initialValue, type, max and properties", () => {
    const input = wrapper.find("input");

    expect(input.prop("value")).toEqual(price);
    expect(input.prop("type")).toEqual("number");
    expect(input.prop("max")).toEqual(INITIAL_ACCOUNTS[0].value);
  });

  it("[input] Should call onChangePrice function after change input component", () => {
    const input = wrapper.find("input");

    expect(onChangePrice).toBeCalledTimes(0);
    input.simulate("change", {
      persist: jest.fn(),
      target: { value: 1 },
    });
    expect(onChangePrice).toBeCalledTimes(1);
    expect(onChangePrice).toHaveBeenCalledWith("source", "1.00");
  });

  it("[select] Should call onChangeAccoun function after change select component", () => {
    const select = wrapper.find(".select");
    expect(onChangeAccount).toBeCalledTimes(0);
    select.simulate("change", {}, { value: TEST_RATES[2].value });
    expect(onChangeAccount).toBeCalledTimes(1);
    expect(onChangeAccount).toHaveBeenCalledWith("source", TEST_RATES[2].value);
  });
});
