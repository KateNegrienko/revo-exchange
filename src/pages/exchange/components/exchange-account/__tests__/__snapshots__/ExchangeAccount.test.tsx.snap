// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Exchange account component] [Exchange account snapshot] Should render properly with only required props  1`] = `
<div
  class="card root"
>
  <div
    class="error"
  />
  <div
    class="row"
  >
    <div
      aria-expanded="false"
      class="ui selection dropdown select"
      name="EUR"
      role="listbox"
      tabindex="0"
    >
      <div
        aria-atomic="true"
        aria-live="polite"
        class="divider text"
        role="alert"
      >
        EUR
      </div>
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="menu transition"
      >
        <div
          aria-checked="false"
          aria-selected="false"
          class="item"
          price="1"
          role="option"
          style="pointer-events:all"
        >
          <span
            class="text"
          >
            USD
          </span>
        </div>
        <div
          aria-checked="true"
          aria-selected="true"
          class="active selected item"
          price="0.82"
          role="option"
          style="pointer-events:all"
        >
          <span
            class="text"
          >
            EUR
          </span>
        </div>
        <div
          aria-checked="false"
          aria-selected="false"
          class="item"
          price="0.73"
          role="option"
          style="pointer-events:all"
        >
          <span
            class="text"
          >
            GBP
          </span>
        </div>
      </div>
    </div>
    <div
      class="ui small left icon input input"
    >
      <i
        aria-hidden="true"
        class="minus icon"
      />
      <input
        max="50.3"
        pattern="^\\\\d*(\\\\.\\\\d{0,2})?$"
        step=".01"
        type="number"
        value=""
      />
    </div>
  </div>
  <div
    class="hint"
  >
     Balance: 50.3
  </div>
</div>
`;


